### PR TITLE
CSJ - Pruned_Transducer_Stateless5

### PR DESCRIPTION
**Pull request by teowenshen**
This is the first recipe that I ran for CSJ after aligning the transcript generation with espnet.

The model is still training now. These are the results at 10 epochs. I will update again next week when it reaches 30 epochs.

| testset | CER |
| -- | -- |
| eval1 | 6.01 |
| eval2 | 4.71 | 
| eval3 | 4.89 | 

Also, although I didn't include a TelegramLogging file which is part of a notification system, I attempt an import of the module at the start of `train.py`. I hope that's okay.

### CODE REVIEW

**Comment by csukuangfj** [comment]
Thanks! The results look promising compared with the one listed in
https://github.com/espnet/espnet/blob/master/egs/csj/asr1/RESULTS.md

By the way, what are the decoding commands you are using?

**Comment by teowenshen** [comment]
At the moment, my train and decode commands are as below. I've included them as separate files in train_cmd.sh and decode_cmd.sh. 

**decoding command**
> ./pruned_transducer_stateless5/decode.py \
>     --num-encoder-layers 18 \
>     --dim-feedforward 2048 \
>     --nhead 8 \
>     --encoder-dim 512 \
>     --decoder-dim 512 \
>     --joiner-dim 512 \
>     --context-size 3 \
>     --simulate-streaming 1 \
>     --decode-chunk-size 16 \
>     --left-context 64 \
>     --causal-convolution 1 \
>     --epoch 10 \
>     --avg 3 \
>     --exp-dir pruned_transducer_stateless5/exp_disfluent \
>     --max-sym-per-frame 1 \
>     --max-duration 155 \
>     --decoding-method modified_beam_search \
>     --beam-size 12  \
>     --word-table lang_char_disfluent/words.txt \
>     --transcript-mode disfluent

**training command**
> python pruned_transducer_stateless5/train.py \
>   --exp-dir pruned_transducer_stateless5/exp_disfluent \
>   --num-encoder-layers 18 \
>   --dim-feedforward 2048 \
>   --nhead 8 \
>   --encoder-dim 512 \
>   --decoder-dim 512 \
>   --joiner-dim 512 \
>   --dynamic-chunk-training 1 \
>   --causal-convolution 1 \
>   --short-chunk-size 20 \
>   --num-left-chunks 4 \
>   --max-duration 155 \
>   --world-size 1 \
>   --start-epoch 11 \
>   --num-epochs 10 \
>   --transcript-mode disfluent \
>   --context-size 3 \
>   --telegram-cred misc.ini \
>   --word-table lang_char_disfluent/words.txt \
>   --musan-dir /mnt/minami_data_server/t2131178/corpus/musan/musan/fbank || \
> python notify_tg.py "Something went wrong during training"

A deviation I made from the Librispeech recipe during training was using a context-size of 3. 
As for decoding, I used quite a large beam-size of 12.

Although I am not familiar with how wenet prepares their transcript, it seemed to have also surpassed their results.
https://github.com/wenet-e2e/wenet/tree/main/examples/csj/s0

**Comment by csukuangfj** [comment]
Thanks! 

Maybe you can try --avg 1 or 2.
3 is a little larger for epoch 10 during decoding.

**Comment by csukuangfj** [comment]
By the way, if you provide several beam sizes for decode.py, please also include the beam size info in the file name of the result file. Otherwise, you will only get one result file for the last beam size value.

**Comment by teowenshen** [comment]
Yes sure. I implemented that in line 697 of decode.py by introducing another `nargs="+"` argument `--beam-sizes`, but I have not tried that yet. Maybe I should streamline all the arguments to align with the other recipes as a final step before merging. 

On the other hand, is there any academic paper you would prefer for us to cite the NGK family? I noticed that Lhotse has its own paper, and k2 has a recent paper on Pruned RNN-T. Please let me know if you'd prefer I open an issue for this inquiry.

**Comment by teowenshen** [comment]
Hi there, I have some weird/interesting findings. 

I tested my model (pruned_transducer_stateless5) at epoch 10, 17, and 24. The CER worsened. 

| Epoch | eval1 | eval2 | eval3 | 
| ------ | ------| ------ | ------ |
| Epoch 10 | 5.64 | 4.61 | 4.74 | 
| Epoch 17 | 7.35 | 5.93 | 6.29 | 
| Epoch 24 | 7.9 | 7.03 | 8.15 |

This round, I am actually reproducing an earlier result where I also tried pruned_transducer_stateless5 and got the same worsening CER. Since then, I have reinstalled the whole NGK package, redone the data preparation stage, and rewritten the training codes. 

Upon closer inspection, I notice the worsening CER is contributed by hallucinated trailing fillers. 
These are a few decoding results at different epochs. I specifically picked these segments because I can confirm there are no audible sounds at all. 

image
えー is pronounced eh.  
image
えっと is pronounced etto. 

These are not the only fillers. I have also double checked the training data - similar to the evaluation data, fillers seldom occur at the end, more often occurring at the start and middle of segments.

Do you face this issue with other recipes? 

I'm using avg=1, beam-size=12 for all 3 setups. Below is my decode command. Just substitute X with the epoch number. 

> ./pruned_transducer_stateless5/decode.py \
>     --num-encoder-layers 18 \
>     --dim-feedforward 2048 \
>     --nhead 8 \
>     --encoder-dim 512 \
>     --decoder-dim 512 \
>     --joiner-dim 512 \
>     --context-size 3 \
>     --simulate-streaming 1 \
>     --decode-chunk-size 16 \
>     --left-context 64 \
>     --causal-convolution 1 \
>     --epoch X \
>     --avg 1 \
>     --exp-dir pruned_transducer_stateless5/exp_disfluent \
>     --max-sym-per-frame 1 \
>     --max-duration 155 \
>     --decoding-method modified_beam_search \
>     --beam-size 12  \
>     --word-table lang_char_disfluent/words.txt \
>     --transcript-mode disfluent

**Comment by csukuangfj** [comment]
Could you disable the tail paddings and retry?

https://github.com/k2-fsa/icefall/blob/884e7bcff3b5cec43705ecb415ae94984b69962d/egs/csj/ASR/pruned_transducer_stateless5/decode.py#L394-L399


Also, for later epochs, I suggest that you try several values of `--avg` during decoding.

**Comment by danpovey** [comment]
There might be some kind of mismatch if we sometimes pad with zeros in test time but never pad with zeros in training time.  (padding a small proportion of training utterances with zeros, i.e. -23 == log(0), might make it a bit more robust to this?)
Also in future, if we ever implement predicting an EOS symbol, that might possibly help as well.

**Comment by teowenshen** [comment]
Yes. The results improved! Now, for epoch 24 (avg3): 2.0 (valid), 5.37, 4.19, 4.28.

These are the results at epoch 27, beam-size 12. 

| Avg | valid | eval1 | eval2 | eval3 | 
| ---- | ----- | ------ | ----- | ----- |
| 1 | 1.91 | 5.50 | 4.22 | 4.41 |
| 3 | **1.89** | 5.42 | 4.18 | 4.28 |
| 5 | 1.91 | 5.42 | 4.19 | 4.33 |
| 6 | 1.92 | 5.38 | 4.16 | 4.22 | 
| **7** | 1.93 | **5.35** | **4.15** | **4.18** | 
| 8 | 1.94 | **5.35** | 4.16 | 4.23 |
| 9 | 1.97 | 5.38 | 4.19 | 4.25 |

However, is there anything we lose if we remove the padding during testing? I understand that removing the padding makes the length shorter after subsampling, but in this case it helped with the CER.

**Comment by csukuangfj** [comment]
> However, is there anything we lose if we remove the padding during testing

The padding is mainly used to avoid deletion errors at the end of the sentence. If you don't have such deletion errors, then I think we lose nothing.

**Comment by teowenshen** [comment]
I see. I wonder why such deletions don't occur in the CSJ case. 

Anyhow, I am continuing the training until Epoch 35, just to see what happens. 
The higher CERs of the later epochs were mostly contributed by (a) misspelled disfluencies (long, short vowels) and (b) wrong homonyms.

**Comment by teowenshen** [comment]
Hi, another update. 

I stopped the training when the valid loss plateaued at epoch 33. However, it would seem that the model overfit. These results are all from epoch 33. 

| avg | beam | valid | eval1 | eval2 | eval3 | 
| ---- | ------ | ---- | ---- | ------ | ---- |
| 1 | 4 | 1.82 | 5.59 | 4.27 | 4.48 | 
| 1 | 12 | 1.82 | 5.6 | 4.25 | 4.44 |  
| 5 | 4 | 1.79 | 5.47 | 4.15 | 4.41 | 
| 5 | 12 | **1.77** | 5.46 | 4.13 | 4.39 | 
| 10 | 4 | 1.82 | 5.45 | 4.16 | 4.43 | 
| 10 | 12 | 1.79 | 5.45 | 4.16 | 4.44 | 
| 15 | 4 | 1.88 |  5.38 | 4.11 | 4.29 | 
| 15 | 12 | 1.87 | 5.38 | **4.10** | 4.28 | 
| **20** | 4 | 2.05 | **5.37** | 4.15 | 4.29 | 
| **20** | 12 | 2.04 | 5.39 | 4.15 | **4.27** | 

The decoding at epoch 27 which yielded 5.35 for eval1 indicates that the overfitting happened much earlier. I am doing a more exhaustive analysis between epoch 25 and epoch 28 to find out the optimal average range. 

While I suspect this might be a corpus/language-related issue, do you face such overfitting issues with other corpora or recipes, even though the trend of the validation loss does not correctly reflect? In CSJ, the validation set is split from shuffled training data.

**Comment by csukuangfj** [comment]
> do you face such overfitting issues with other corpora or recipes

Yes, We have the same issue for other dataset, e.g., librispeech and wenetspeech.

**Comment by csukuangfj** [comment]
> Should I close this pull request? I am thinking to refork, since the history of this branch is quite messy.

@danpovey is going to release an optimized model soon. Maybe you can use that model for training. It is much better and
also faster to train.

**Comment by teowenshen** [comment]
That sounds exciting. Sure. 

I will close this pull request now.
